frontend.md

```diff
@@ -12,20 +12,65 @@ A few general rules for frontend:
- - **`HypixelPlayer`** **must** implement the following methods:
-    - A method that returns the friends of the player in the form of a list of **`HypixelPlayer`** data structures.
-    - A method that returns the online status of the player in the form of a **`HypixelPlayerStatus`** data structure.
-    - A method that returns the recent games of the player in the form of a list of **`HypixelGameSession`** data structures.
-    - A method that returns either a **`HypixelGuild`** if the player is in a guild or nothing if the player is not in a guild.</br>
+**Note**: When reading the information below, multiple methods that return the same data **must** only ever be allowed because of multiple independent arguments.
+
+**Note**: This information explicitly does not describe how the different data structures should behave internally or what data they need to store. This is reserved for the backend and should not concern the frontend enough to be listed here.
+
+- A main data structure called `HypixelAPI` **must** exist in the wrapper. `HypixelAPI` **must** store the following data:
+
+| name | type | note |
+| --- | --- | --- |
+| API Key | **`UUID`** | This is a key retrieved from Hypixel and provided by the developer. |
+| Cache | **`HypixelCache`** | This may be a reference. `HypixelCache` is an interface, see below. |
+| Cache Strategy | **`HypixelCacheStrategy`** | This may be a reference. `HypixelCacheStrategy` is an interface, see below. |
+
+- **`HypixelAPI` must** implement one or more methods for every row in the following table:
+
+| arguments | return type | note |
+| --- | --- | --- |
+| or `UUID` of the player</br>or `Name` of the player | **`HypixelPlayer`** | `Name` is a string. |
+| or `UUID` of player in guild</br>or `ID` of requested guild</br>or `Name` of the requested guild | **`HypixelGuild`** | `Name` is a string. |
+| | **`HypixelLeaderboards`** |
+| | **`HypixelPunishmentStats`** |
+
+**Note**: If a language's design and conventions permit it, these methods may be structurally grouped together, branching off of `HypixelAPI`.
+
+- **`HypixelPlayer`** **must** implement the following methods:
+
+| arguments | return type | note |
+| --- | --- | --- |
+| | **`HypixelPlayerStatus`** | This returns the online status of the player. |
+| | collection of **`HypixelFriendShip`** | This returns the friends of the player. |
+| | collection of **`HypixelGameSession`** | This returns the recent games of the player. |
+| | optional **`HypixelGuild`** | This returns guild of the player. |
+
+- **`HypixelPlayerStatus`** **must** implement the following methods:
+
+| arguments | return type | note |
+| --- | --- | --- |
+| | **`boolean`** | This returns whether the player is currently online. |
+| | **`GameType`** | This returns the game the player is in. |
+| | **`String`** | This returns the mode of the game. |
+| | **`String`** | This returns the map of the game. |
+
+- **`HypixelFriendShip`** **must** implement the following methods:
```

Comment:
Friendship is 1 word, so the "S" should be lowercase
```suggestion
- **`HypixelFriendship`** **must** implement the following methods:
```
This also applies to line 43